### DISCUSSION TITLE
How to handle multiple iOS targets (widgets) in fastlane

### DISCUSSION THREAD

**Discussion by ir2pid** [fastlane/fastlane#17378] [Q&A]
I'm trying to add widgets to our project and fastlane isn't able to automate anymore.

- How do I edit the widget bundle identifier in fastlane
- How do I set the right provisioning profile for the right target

Things I need:

- in xcode the app target should have bundle identifier x.y.z
- in xcode the widget target should have the bundle identifier x.y.z.widget

app target
<img width="696" alt="Screenshot 2020-10-05 at 10 10 15" src="https://user-images.githubusercontent.com/907075/95070812-f5ce4180-0708-11eb-8a3a-f80e17c8e5cd.png">

widget target
<img width="661" alt="Screenshot 2020-10-05 at 10 10 39" src="https://user-images.githubusercontent.com/907075/95070705-cd464780-0708-11eb-83ab-5b816a1c1319.png">

- in xcode the app target should have provisioning profile match adhoc x.y.z
- in xcode the widget target should have provisioning profile match adhoc x.y.z.widget

app target signing
<img width="694" alt="Screenshot 2020-10-05 at 10 10 26" src="https://user-images.githubusercontent.com/907075/95070755-ddf6bd80-0708-11eb-9793-2501195c0951.png">
widget target signing
<img width="698" alt="Screenshot 2020-10-05 at 10 10 48" src="https://user-images.githubusercontent.com/907075/95070787-ea7b1600-0708-11eb-8e22-d5a190feee1f.png">

current Fastfile lane:
```

  desc "Build the app and send it to Testflight for testing"
  lane :build_adhoc do
    UI.message("app_name: #{app_name}")
    UI.message("app_identifier: #{app_identifier}")
    UI.message("apple_id: #{apple_id}")
    UI.message("team_id: #{team_id}")
    UI.message("sku: #{sku}")

    xcodeprojpath = "../ios/" + app_name + ".xcodeproj"

    proj = Xcodeproj::Project.open("../" + xcodeprojpath)

    proj.build_configurations.each do |item|
      item.build_settings["DEVELOPMENT_TEAM"] = team_id
      item.build_settings["PROVISIONING_PROFILE_SPECIFIER"] = match_ad_hoc_provisioning
      item.build_settings["CODE_SIGN_IDENTITY[sdk=iphoneos*]"] = match_ad_hoc_signing
      item.build_settings["SWIFT_VERSION"] = swift_version
      item.build_settings["ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES"] = "YES"
    end

    proj.recreate_user_schemes
    proj.save

    begin
      disable_automatic_code_signing
    rescue => ex
      UI.message("failed to disable automatic signing")
      UI.error(ex)
    end
    get_certificates(
      development: false,
      username: apple_id,
      team_id: team_id,
    )
    unlock_keychain(
      password: keychainPassword,
    )

    #bumpBuildNumber

    sigh(username: apple_id, adhoc: true, readonly: false, app_identifier: app_identifier, team_id: team_id)  #force: false,
    gym(export_method: "ad-hoc",
        clean: true,
        configuration: "Release",
        codesigning_identity: match_ad_hoc_signing,
        export_team_id: team_id,
        skip_profile_detection: true,
        export_options: {
          method: "ad-hoc",
          signingStyle: "manual",
          provisioningProfiles: { "#{app_identifier}": "#{match_ad_hoc_provisioning}" },
        })
  end
```

**Comment by rogerluan** [comment]
Hi @ir2pid ! 

We have several targets in our project, maybe I can help you with this.
I think you're having difficulties setting up the bundle identifiers of your targets. First, bear in mind that this should be done just once when creating the extension and not in every build, so I think it's not necessarily something that should be part of your fastlane workflow 🤔 Nonetheless, you can achieve that setup by creating `User Defined` env variables, in Xcode:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/8419048/95126695-3f308800-072d-11eb-85aa-fc70842b0773.png">

Then you can create a set of variables that will compound your environment variables. E.g.:

<img width="187" alt="image" src="https://user-images.githubusercontent.com/8419048/95126788-57080c00-072d-11eb-84a2-4b2bc8521afa.png">


Each of those envs can have its own value. Or you can do e.g. `com.apple.project${ANOTHER_USER_DEFINED_ENV_VAR}`, and have `ANOTHER_USER_DEFINED_ENV_VAR` have different values based on the environment, e.g. `.debug`, `.release`, `.sandbox`, etc.

That should take care of your bundle identifiers. In the "Bundle Identifier" you can enter `${BUNDLE_IDENTIFIER}` (name used in the example above), and it will display something like this:

<img width="230" alt="image" src="https://user-images.githubusercontent.com/8419048/95127003-b239fe80-072d-11eb-9fb0-96e9faa12eac.png">

Same applies to code signing and provisioning profiles.

You will have to create these user-defined env vars in the Build Settings of each of your targets, individually. Although this sounds like too much work, it's something you only need to set up once, when you create that extension, so you'll end up doing that like what... once or twice a year? Assuming your app implements all the extensions out there? 😄 

Let me know if that works for you, or if you have any further questions 😊 happy to help! 
Good luck! 💪
